### ISSUE TITLE
The live poller not updating currently running jobs

### ISSUE THREAD

**Issue by jonmallVu** [resque/resque#1775]
Hi,

I recently updated our sinatra project from Resque 1.27.4 to 2.1.0 and the live poller seems to randomly stop updating the currently running jobs. The `Queues` values are sometimes still updated but this is not consistent. For some reason it also seems that the workers are being cleared, after some idle time. This was not the case with 1.27.4.

Have any defaults changed? Any ideas on how I can debug further? 

Thanks in advance.

**Environment:**
Redis (Redis server v=6.2.6) and app hosted on kubernetes as a stateful set and deployment respectively.
Ruby 2.7.4
```ruby
gem 'sinatra', '~> 2.1'
gem 'redis', '~> 4.5.1'
gem 'resque', '~> 2.1.0'
gem 'resque-retry', '~> 1.7.0'
gem 'resque-scheduler', '~> 4.5.0'
```

**api/config.ru**

```ruby
if %w(test development).include? ENV['RACK_ENV']
  require 'dotenv'
  Dotenv.load
end

require 'envied'
ENVied.require

require 'resque/server'
require 'resque-scheduler'
require 'resque/scheduler/server'
require 'resque-retry'
require 'resque-retry/server'
require 'resque/failure/redis'
require_relative '../lib/connect'
require_relative 'app'

Postgres.connect

api = Rack::Builder.new do
  map '/' do
    run VuworkflowApi.new
  end

  map '/status' do
    Resque.redis = ENV['REDIS_SERVER']
    run Resque::Server.new
  end
end


run api
```

startup command `bundle exec puma -b tcp://0.0.0.0:9292 api/config.ru`

**Comment by iloveitaly** [comment]
@jonmallVu thanks for the report! Here are two places to look:

[This change](https://github.com/resque/resque/commit/be71bd1a3e10138284a259e496d7ae01da820646) modified the polling javascript.

[This change](https://github.com/resque/resque/commit/e919fba929cf1df09c3fb149870aea9b3b7a0078) modified how workers are pruned automatically. How long are you seeing works go idle before they are removed? What problems is this causing you?

**Comment by jonmallVu** [comment]
HI @iloveitaly ,

Thanks for your response. It seems that the workers are pruned after about 5-6 minutes of being idle. Just to clarify, the worker containers themselves are still alive and nothing is changing with them, the just stop appearing in the workers list in the Resque UI. 

They do not re-appear when new jobs are added the their queues (after they've been pruned), but the jobs are still processed. The main problem this causes, is we can't see whether a job is hanging or whether it is being executed since we get no feedback from the resque poller. I've attached a short video showing jobs being added and cleared from queues (as they are executed) but the poller not showing any updates of this happening and the workers not appearing in the list of workers. At the end it should be showing that a `Work` worker is executing a job but it doesn't show anything.

Do I need to add code to the workers to send heartbeats to the Resque service? I thought Resque would poll the workers to check for a pulse.

https://user-images.githubusercontent.com/38257327/139659325-3b0cbc7f-2ab3-4766-a3fa-fd75b571b77e.mov

**Comment by iloveitaly** [comment]
@jonmallVu do you see anything in the web console relating to the polling? Is the browser hitting the backend at all?

@mtsmfm do you have any thoughts on why the workers are being pruned?

**Comment by mtsmfm** [comment]
@iloveitaly Hmm, I have no idea why it happens.

@jonmallVu Can you reproduce the problem with resque 2.0.0?

**Comment by mtsmfm** [comment]
I can't reproduce this issue.

```ruby
require 'bundler/inline'

gemfile do
  source 'https://rubygems.org'

  gem 'resque', '2.1.0'
  gem 'sinatra', '2.1.0'
  gem 'webrick'
  gem 'rake'
end

class MyApp < Sinatra::Base 
  get '/' do
    Resque.enqueue(MyJob)

    'Enqueued MyJob'
  end
end

class MyJob
  @queue = :my_job

  def self.perform(*)
    sleep 10
  end
end

api = Rack::Builder.new do
  map '/' do
    run MyApp.new
  end

  map '/status' do
    run Resque::Server.new
  end
end

if ENV['API'] == '1'
  require 'resque/server'
  Rack::Server.start(app: api)
else
  require 'rake'
  require 'resque/tasks'
  ENV['QUEUE'] = '*'
  Rake.application['resque:work'].invoke
end

# 1. Save this file as test.rb
# 2. Run `API=1 ruby test.rb`
# 3. Open another terminal and run `ruby test.rb`
# 4. Open localhost:8080/status and make sure 1 worker available
# 5. Wait more than 5 min
# 6. Access localhost:8080
# 7. Open localhost:8080/status again and make sure 1 job is processed
```

![resque](https://user-images.githubusercontent.com/1796864/139687989-6fa58e0d-f03e-4e2b-a93c-3a64e63f1635.gif)

**Comment by jonmallVu** [comment]
I'm experiencing the same issues with 2.0.0. The workers are pruned after 5 to 6 minutes. I can see the poller triggering every 2 seconds (screenshot attached). @mtsmfm , I'll give it a go and get back to you.

I'm going to try the following two things on my solution:
1. Downgrade back down to 1.27.4 and see if the issue persists. If this is the case I apologise as it must be something else causing the issue.
2. Monitor the polling for the full 6 minutes and inspect the overview.poll requests when the workers are pruned.

Could it be related to the resque-retry and/or resque-scheduler add-ons?


![image](https://user-images.githubusercontent.com/38257327/139710951-81688f33-273a-4d02-a888-24add9f09049.png)

**Comment by iloveitaly** [comment]
[This change](https://github.com/resque/resque/commit/e919fba929cf1df09c3fb149870aea9b3b7a0078) is the only thing to my knowledge that changes how workers are purged. I bet this is the culprit, but I'm not sure what could be causing the issue there.

**Comment by mtsmfm** [comment]
@iloveitaly 

> I bet this is the culprit

No. As @jonmallVu described, the problem still happens even with Resque 2.0.0 and the change is included since 2.1.0.

![image](https://user-images.githubusercontent.com/1796864/139776975-1b67387b-9ef2-44ff-b95f-d84072c7adfc.png)

**Comment by mtsmfm** [comment]
@jonmallVu If you can't reproduce the problem with 1.27.4, you can use `git bisect` to find the commit which causes the problem.

```
$ git clone https://github.com/resque/resque
$ cd resque
$ git bisect start
$ git bisect bad v2.0.0
$ git bisect good v1.27.4
```

- Edit your Gemfile to use the version bisect indicates (like `gem "resque", git: "https://github.com/resque/resque", ref: "5aa672c100ac5ebc3b8be8a6224f55eb0c6a17e3"`
- Make sure the problem happens or not
  - If it still happens, run `git bisect bad` on resque dir and update `ref` in Gemfile again as bisect indicates
  - If it doesn't happen, run `git bisect good` on resque dir and update `ref` in Gemfile again as bisect indicates
- Finally you'll get the revision which causes this problem

**Comment by iloveitaly** [comment]
> No. As @jonmallVu described, the problem still happens even with Resque 2.0.0 and the change is included since 2.1.0.

Ah, interesting. I don't have any strong ideas of what would cause this to occur. 

I took a look through the git compare and this line seems suspect:

https://github.com/resque/resque/compare/v1.27.4...v2.0.0#diff-a43b4a5dd5107cefd407ce6da9c354cbf3c466f1e422df6f96efb003ac84b9b7R595

**Comment by jonmallVu** [comment]
Hi @mtsmfm  @iloveitaly ,

I think the issue is related to my setup. I ran the test posted above and didn't see the same issue. I then deployed my docker image on a different environment and couldn't replicate the issue either. 

Thanks for all your help!
